File: src/server/quant/MarketContext.ts
/**
 * ==========================================================
 * Module: MarketContext
 *
 * Purpose:
 * Phase 3 of the additive quant layer - SPY/QQQ/IWM trend, sector strength, and stock-vs-
 * benchmark/sector relative strength, using the SAME real historical data source everything else
 * in this codebase already uses (HistoricalDataGateway -> ohlcv_bars, Alpaca-backed) - no new
 * data provider is introduced. Every field in the returned object is explicitly source-tagged.
 *
 * Breadth metrics (advance/decline, new highs/lows, etc.) are NOT computed - no data source for
 * market-wide breadth exists anywhere in this codebase, and this module reports that honestly
 * (`breadth.available: false`) rather than fabricating one from the handful of symbols it happens
 * to fetch, per this plan's explicit "do not invent unavailable data" requirement.
 *
 * Sector proxy ETFs (XLK/XLF/etc.) are new to this module - PositionSizing.ts's existing
 * `SECTOR_MAP`/`getSector` (symbol -> sector NAME, used for its own concentration-cap gate) is
 * imported and reused as-is to look up a symbol's sector; this file only adds the NEW mapping
 * from sector name -> a representative sector ETF ticker, it does not touch PositionSizing.ts.
 * ==========================================================
 */
import { Bar, historicalDataGateway } from '../engines/backtest/HistoricalDataGateway';
import { getSector } from '../engines/PositionSizing';
import { classifyRegime, RegimeResult } from './RegimeEngine';
import { correlation, beta } from './statistics';

export const BENCHMARK_SYMBOLS = { broad: 'SPY', tech: 'QQQ', smallCap: 'IWM' } as const;

/** Representative sector ETF per GICS-ish sector name - matches PositionSizing.ts's own sector
 *  naming exactly (both must agree for `getSector(symbol)`'s result to look up correctly here). */
export const SECTOR_ETF_MAP: Record<string, string> = {
  'Technology': 'XLK',
  'Communication Services': 'XLC',
  'Consumer Discretionary': 'XLY',
  'Financials': 'XLF',
  'Energy': 'XLE',
  'Healthcare': 'XLV',
  'Consumer Staples': 'XLP',
  'Industrials': 'XLI',
};

export interface BenchmarkTrend {
  symbol: string;
  regime: RegimeResult | null; // null only if bars genuinely couldn't be fetched
  source: string;
}

export interface RelativeStrength {
  vsSymbol: string;
  periodPct: number | null; // symbol's own % change over the lookback window
  benchmarkPeriodPct: number | null;
  relativeStrengthPct: number | null; // symbol's periodPct minus the benchmark's - real outperformance/underperformance, not a fabricated score
  correlation: number | null;
  beta: number | null;
  source: string;
}

export interface MarketContextResult {
  spy: BenchmarkTrend;
  qqq: BenchmarkTrend;
  iwm: BenchmarkTrend;
  sector: { name: string | null; etf: string | null; trend: BenchmarkTrend | null };
  relativeStrengthVsSPY: RelativeStrength | null;
  relativeStrengthVsSector: RelativeStrength | null;
  breadth: { available: false; reason: string };
}

/** Injectable so tests never need a real Alpaca-backed DB fetch - production code (getMarketContext's
 *  default) goes through the real historicalDataGateway, identical to how BacktestEngine/RiskEngine
 *  already source real bars. */
export type BarsFetcher = (symbol: string, timeframe: string, startMs: number, endMs: number) => Promise<Bar[]>;

export const defaultBarsFetcher: BarsFetcher = async (symbol, timeframe, startMs, endMs) => {
  await historicalDataGateway.ensureBars(symbol, timeframe, startMs, endMs);
  return historicalDataGateway.getBars(symbol, timeframe, startMs, endMs);
};

async function benchmarkTrend(symbol: string, timeframe: string, startMs: number, endMs: number, fetchBars: BarsFetcher): Promise<BenchmarkTrend> {
  try {
    const bars = await fetchBars(symbol, timeframe, startMs, endMs);
    if (bars.length === 0) return { symbol, regime: null, source: `ohlcv_bars(alpaca):${symbol}:${timeframe} - no bars returned` };
    return { symbol, regime: classifyRegime(bars), source: `ohlcv_bars(alpaca):${symbol}:${timeframe}` };
  } catch (e: any) {
    return { symbol, regime: null, source: `ohlcv_bars(alpaca):${symbol}:${timeframe} - fetch failed: ${e.message}` };
  }
}

function pctChange(bars: Bar[]): number | null {
  if (bars.length < 2) return null;
  const first = bars[0].close;
  if (first === 0) return null;
  return ((bars[bars.length - 1].close - first) / first) * 100;
}

async function computeRelativeStrength(symbolBars: Bar[], benchmarkSymbol: string, timeframe: string, startMs: number, endMs: number, fetchBars: BarsFetcher): Promise<RelativeStrength | null> {
  try {
    const benchmarkBars = await fetchBars(benchmarkSymbol, timeframe, startMs, endMs);
    if (benchmarkBars.length === 0 || symbolBars.length === 0) return null;

    const periodPct = pctChange(symbolBars);
    const benchmarkPeriodPct = pctChange(benchmarkBars);
    const relativeStrengthPct = periodPct !== null && benchmarkPeriodPct !== null ? periodPct - benchmarkPeriodPct : null;

    const symbolCloses = symbolBars.map(b => b.close);
    const benchCloses = benchmarkBars.map(b => b.close);

    return {
      vsSymbol: benchmarkSymbol,
      periodPct,
      benchmarkPeriodPct,
      relativeStrengthPct,
      correlation: correlation(symbolCloses, benchCloses),
      beta: beta(symbolCloses, benchCloses),
      source: `ohlcv_bars(alpaca):${benchmarkSymbol}:${timeframe}`,
    };
  } catch {
    return null;
  }
}

/**
 * Full market-context assessment for `symbol` over `[startMs,endMs]`. `symbolBars` is passed in
 * (the caller - QuantSignalAgent - already fetched it for its own indicator computation; this
 * avoids a redundant fetch of the same symbol/range).
 */
export async function getMarketContext(
  symbol: string,
  symbolBars: Bar[],
  timeframe: string,
  startMs: number,
  endMs: number,
  fetchBars: BarsFetcher = defaultBarsFetcher
): Promise<MarketContextResult> {
  const [spy, qqq, iwm] = await Promise.all([
    benchmarkTrend(BENCHMARK_SYMBOLS.broad, timeframe, startMs, endMs, fetchBars),
    benchmarkTrend(BENCHMARK_SYMBOLS.tech, timeframe, startMs, endMs, fetchBars),
    benchmarkTrend(BENCHMARK_SYMBOLS.smallCap, timeframe, startMs, endMs, fetchBars),
  ]);

  const sectorName = getSector(symbol);
  const sectorEtf = sectorName ? SECTOR_ETF_MAP[sectorName] ?? null : null;
  const sectorTrend = sectorEtf ? await benchmarkTrend(sectorEtf, timeframe, startMs, endMs, fetchBars) : null;

  const [relativeStrengthVsSPY, relativeStrengthVsSector] = await Promise.all([
    computeRelativeStrength(symbolBars, BENCHMARK_SYMBOLS.broad, timeframe, startMs, endMs, fetchBars),
    sectorEtf ? computeRelativeStrength(symbolBars, sectorEtf, timeframe, startMs, endMs, fetchBars) : Promise.resolve(null),
  ]);

  return {
    spy, qqq, iwm,
    sector: { name: sectorName, etf: sectorEtf, trend: sectorTrend },
    relativeStrengthVsSPY,
    relativeStrengthVsSector,
    breadth: { available: false, reason: 'No market-breadth data source (advance/decline, new highs/lows) exists in this codebase - not fabricated from the handful of symbols fetched here.' },
  };
}

File: quant-core-java/src/main/java/io/argus/quantcore/features/MarketContext.java
package io.argus.quantcore.features;

import io.argus.quantcore.backtest.engine.Bar;

import java.util.ArrayList;
import java.util.List;

/**
 * Ported from src/server/quant/MarketContext.ts's getMarketContext (JMIG-001) - restricted to the
 * PURE computational core. The real TS function's bar-fetching is already injectable via its own
 * {@code BarsFetcher} type specifically so callers/tests can supply already-fetched bars instead
 * of a live network/DB call (see MarketContext.test.ts's own {@code fakeFetcher}); this port takes
 * that same shape one step further and requires every bar array as a plain input, doing no
 * fetching of its own at all - consistent with quant-core-java being a computational engine with
 * zero broker/DB/network access of its own.
 *
 * Sector name/ETF resolution (PositionSizing.ts's {@code getSector}/{@code SECTOR_MAP}) is
 * likewise a caller-supplied input here ({@code sectorName}/{@code sectorEtf}), not ported - out
 * of JMIG-001's stated file scope (RegimeEngine/MarketContext/trend/volatility/priceAction/volume/
 * supportResistance only; PositionSizing.ts is a different file).
 */
public final class MarketContext {
    private MarketContext() {
    }

    /** Models one benchmark's real fetch outcome: {@code bars} populated (success, possibly
     *  empty), or {@code fetchError} populated (the real TS try/catch's error message) - mirrors
     *  getMarketContext's two real failure branches ("no bars returned" vs "fetch failed: ..."). */
    public record BenchmarkInput(List<Bar> bars, String fetchError) {
        public static BenchmarkInput ok(List<Bar> bars) {
            return new BenchmarkInput(bars, null);
        }

        public static BenchmarkInput failed(String error) {
            return new BenchmarkInput(null, error);
        }
    }

    public record BenchmarkTrend(String symbol, RegimeEngine.Result regime, String source) {
    }

    public record RelativeStrength(String vsSymbol, Double periodPct, Double benchmarkPeriodPct,
                                    Double relativeStrengthPct, Double correlation, Double beta, String source) {
    }

    public record Sector(String name, String etf, BenchmarkTrend trend) {
    }

    public record Breadth(boolean available, String reason) {
    }

    public record Result(BenchmarkTrend spy, BenchmarkTrend qqq, BenchmarkTrend iwm, Sector sector,
                          RelativeStrength relativeStrengthVsSPY, RelativeStrength relativeStrengthVsSector,
                          Breadth breadth) {
    }

    static BenchmarkTrend benchmarkTrend(String symbol, String timeframe, BenchmarkInput input) {
        if (input.fetchError() != null) {
            return new BenchmarkTrend(symbol, null,
                "ohlcv_bars(alpaca):" + symbol + ":" + timeframe + " - fetch failed: " + input.fetchError());
        }
        if (input.bars() == null || input.bars().isEmpty()) {
            return new BenchmarkTrend(symbol, null, "ohlcv_bars(alpaca):" + symbol + ":" + timeframe + " - no bars returned");
        }
        return new BenchmarkTrend(symbol, RegimeEngine.classifyRegime(input.bars()), "ohlcv_bars(alpaca):" + symbol + ":" + timeframe);
    }

    static Double pctChange(List<Bar> bars) {
        if (bars.size() < 2) {
            return null;
        }
        double first = bars.get(0).close();
        if (first == 0) {
            return null;
        }
        return ((bars.get(bars.size() - 1).close() - first) / first) * 100;
    }

    static double[] toReturns(List<Bar> bars) {
        List<Double> rets = new ArrayList<>();
        for (int i = 1; i < bars.size(); i++) {
            double prevClose = bars.get(i - 1).close();
            if (prevClose != 0) {
                rets.add(bars.get(i).close() / prevClose - 1);
            }
        }
        double[] arr = new double[rets.size()];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = rets.get(i);
        }
        return arr;
    }

    static RelativeStrength computeRelativeStrength(List<Bar> symbolBars, String benchmarkSymbol, String timeframe, BenchmarkInput benchmarkInput) {
        if (benchmarkInput.fetchError() != null || benchmarkInput.bars() == null
            || benchmarkInput.bars().isEmpty() || symbolBars.isEmpty()) {
            return null;
        }
        List<Bar> benchmarkBars = benchmarkInput.bars();

        Double periodPct = pctChange(symbolBars);
        Double benchmarkPeriodPct = pctChange(benchmarkBars);
        Double relativeStrengthPct = (periodPct != null && benchmarkPeriodPct != null) ? periodPct - benchmarkPeriodPct : null;

        double[] symbolReturns = toReturns(symbolBars);
        double[] benchReturns = toReturns(benchmarkBars);

        return new RelativeStrength(benchmarkSymbol, periodPct, benchmarkPeriodPct, relativeStrengthPct,
            StatisticsMath.correlation(symbolReturns, benchReturns, 20),
            StatisticsMath.beta(symbolReturns, benchReturns, 20),
            "ohlcv_bars(alpaca):" + benchmarkSymbol + ":" + timeframe);
    }

    /** Pure-core equivalent of getMarketContext - see this class's header for what's deliberately
     *  caller-supplied instead of fetched internally. */
    public static Result getMarketContext(List<Bar> symbolBars, String timeframe,
                                           BenchmarkInput spy, BenchmarkInput qqq, BenchmarkInput iwm,
                                           String sectorName, String sectorEtf, BenchmarkInput sector) {
        BenchmarkTrend spyTrend = benchmarkTrend("SPY", timeframe, spy);
        BenchmarkTrend qqqTrend = benchmarkTrend("QQQ", timeframe, qqq);
        BenchmarkTrend iwmTrend = benchmarkTrend("IWM", timeframe, iwm);

        BenchmarkTrend sectorTrend = sectorEtf != null ? benchmarkTrend(sectorEtf, timeframe, sector) : null;

        RelativeStrength relativeStrengthVsSPY = computeRelativeStrength(symbolBars, "SPY", timeframe, spy);
        RelativeStrength relativeStrengthVsSector = sectorEtf != null
            ? computeRelativeStrength(symbolBars, sectorEtf, timeframe, sector)
            : null;

        return new Result(spyTrend, qqqTrend, iwmTrend,
            new Sector(sectorName, sectorEtf, sectorTrend),
            relativeStrengthVsSPY, relativeStrengthVsSector,
            new Breadth(false, "No market-breadth data source (advance/decline, new highs/lows) exists in this codebase - not fabricated from the handful of symbols fetched here."));
    }
}
